### PR TITLE
Overlay

### PR DESCRIPTION
**Pull request by jbeta51**
Now plays video with both user skeletons and template skeletons. Easy to modify if we decide to change the approach of the overlay video. #9

### CODE REVIEW

File: stance/motion/squat.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 import logging
 import json
 
+import matplotlib
+matplotlib.use('TkAgg')
```
Comment (RFinkelberg) [inline stance/motion/squat.py]: This is a MacOS specific fix, shouldn't be here

File: stance/predict/overlay.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
 import cv2
+import matplotlib
+matplotlib.use('TkAgg')
```
Comment (RFinkelberg) [inline stance/predict/overlay.py]: Same comment as above

File: stance/predict/overlay.py [hunk 3]
```diff
@@ -13,7 +16,51 @@ def display_overlay(input_video, template_skeletons):
         path to the video of the user performing the motion
     template_skeletons : List[Skeletons]
     """
-    pass
+    cap = cv2.VideoCapture(input_video)
+    numFrames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
```
Comment (RFinkelberg) [inline stance/predict/overlay.py:21]: Instead of this, consider using a `while success` the same way @aspadiyath  did in `etl.py`

File: stance/predict/overlay.py [hunk 4]
```diff
@@ -13,7 +16,51 @@ def display_overlay(input_video, template_skeletons):
...
+    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+
+    # used for scaling template skeleton to user video
+    scale_x = width / 1280
```
Comment (RFinkelberg) [inline stance/predict/overlay.py:27]: Where are 1280 and 720 coming from?

File: stance/main.py [hunk 5]
```diff
@@ -58,7 +58,8 @@ def main():
 
     # Overlay the template skeletons on the user's video
     logger.info("Overlaying Video")
-    overlay.display_overlay(args.video_path, motion.template_skeletons)
+    print(args.video_path, motion.template_skeletons, benchmark_zone_indices)
```
Comment (RFinkelberg) [inline stance/main.py]: Did you mean to take out this print? If not, use the logger

File: stance/predict/overlay.py [hunk 6]
```diff
@@ -13,7 +16,51 @@ def display_overlay(input_video, template_skeletons):
...
+            dif_y = temp_center[1] * scale_y - user_center[1]
+            for points in temp:
+                if points is not None:
+                    pointA = (int(points.head[0] * scale_x - dif_x), int(points.head[1] * scale_y - dif_y))
+                    pointB = (int(points.tail[0] * scale_x - dif_x), int(points.tail[1] * scale_y - dif_y))
+                    cv2.line(frame, pointA, pointB, (0, 255, 0), 4)
+        if frame is not None:
+            cv2.imshow('Frame', frame)
+        # pause video when skeleton is visible
+        if i == benchmark_indices[count - 1] + 1:
```
Comment (RFinkelberg) [inline stance/predict/overlay.py:63]: If we're worried that the skeleton isn't on there long enough to be visible, should we instead wait for a keypress to move on?

File: stance/predict/overlay.py [hunk 7]
```diff
@@ -13,7 +16,51 @@ def display_overlay(input_video, template_skeletons):
...
+            time.sleep(1)
+        if cv2.waitKey(70) & 0xFF == ord('q'):
+            break
+
+    cap.release()
+    cv2.destroyAllWindows()
 
 
 def compare_user_skeletons(input_video, idxs, user_skeletons, motion):
```
Comment (RFinkelberg) [inline stance/predict/overlay.py:72]: I know its not part of this PR, but is this code deprecated @aspadiyath ?

**Review by RFinkelberg — DISMISSED** [review]
Looks good (assuming it works) apart from some minor stuff. I want @aspadiyath to look at it too

File: stance/predict/overlay.py (hunk 7 above)
Comment (aspadiyath) [reply stance/predict/overlay.py:72]: No, it does a different task. I just put it in overlay for simplicity.

File: stance/predict/overlay.py (hunk 4 above)
Comment (aspadiyath) [reply stance/predict/overlay.py]: It looks like you've hard-coded it for the squat.mp4, when I run your code with the other two examples, it either shows it too far to the right, or not at all.

File: stance/predict/overlay.py [hunk 8]
```diff
@@ -1,8 +1,11 @@
 import cv2
+import matplotlib
+matplotlib.use('TkAgg')
 import matplotlib.pyplot as plt
+import time
+import numpy as np
```
Comment (aspadiyath) [inline stance/predict/overlay.py]: I don't see you using numpy anywhere? you can delete this

File: stance/predict/overlay.py [hunk 9]
```diff
@@ -1,8 +1,11 @@
...
 
-
-def display_overlay(input_video, template_skeletons):
+def display_overlay(input_video, template_skeletons, benchmark_indices, user_skeletons):
```
Comment (aspadiyath) [inline stance/predict/overlay.py:5]: Add everything in here to the docstring

File: stance/predict/overlay.py [hunk 10]
```diff
@@ -13,7 +16,51 @@ def display_overlay(input_video, template_skeletons):
...
+    # used for scaling template skeleton to user video
+    scale_x = width / 1280
+    scale_y = height / 720
+    count = 0
+
+    for i in range(numFrames):
+        _, frame = cap.read()
+
+        # User skeleton
+        temp = list(user_skeletons[i].vectors.values())
```
Comment (aspadiyath) [inline stance/predict/overlay.py]: When I run the examples, this keeps throwing an index out of bounds?

Comment (aspadiyath) [reply stance/predict/overlay.py]: You could also inline this since you don't use it again, and then you have a different "temp" variable later

File: stance/predict/overlay.py (hunk 6 above)
Comment (aspadiyath) [reply stance/predict/overlay.py:63]: I think it's fine, if we save the video to a file. Jorge you can look at what I did [here](https://github.com/RFinkelberg/Stance/blob/develop/stance/main.py#L39) to save it to a file. Maybe put a command line argument for the option to save the output to a file.

**Review by aspadiyath — CHANGES_REQUESTED** [review]
Nice job son, just a couple changes. I love you.

File: stance/predict/overlay.py (hunk 6 above)
Comment (jbeta51) [reply stance/predict/overlay.py:63]: maybe just always have it save? no point in not when using the command

File: stance/predict/overlay.py (hunk 8 above)
Comment (jbeta51) [reply stance/predict/overlay.py]: I'm gonna do a pass through for the entire project. There are other files that have unused imports as well

File: stance/predict/overlay.py (hunk 6 above)
Comment (aspadiyath) [reply stance/predict/overlay.py:63]: Sure, we can always change it if we see fit later

File: stance/predict/overlay.py (hunk 4 above)
Comment (jbeta51) [reply stance/predict/overlay.py:27]: The dimension space we derived the template skeleton from. We could save it as a global variable and retrieve it from there, but for the sake of this assignment I think this is fine.

File: stance/predict/overlay.py (hunk 6 above)
Comment (jbeta51) [reply stance/predict/overlay.py:63]: I'm going to do the video saving edits under a different branch after we merge this one for the sake of organization

File: stance/predict/overlay.py (hunk 10 above)
Comment (jbeta51) [reply stance/predict/overlay.py]: Meant to push the fixed version. My bad. Also I added a try except that would make it awkward to do inline

File: stance/motion/squat.py [hunk 11]
```diff
@@ -3,6 +3,7 @@
 import logging
 import json
 
+import matplotlib
```
Comment (aspadiyath) [inline stance/motion/squat.py]: Why is this here?

File: stance/predict/overlay.py [hunk 12]
```diff
@@ -12,8 +14,59 @@ def display_overlay(input_video, template_skeletons):
     input_video : String
         path to the video of the user performing the motion
     template_skeletons : List[Skeletons]
+    benchmark_indices : frames where benchmark skeletons should be overlayed
```
Comment (aspadiyath) [inline stance/predict/overlay.py]: You need the type as well. It should be like 

variable : Type
\t description

File: stance/predict/overlay.py (hunk 3 above)
Comment (aspadiyath) [reply stance/predict/overlay.py:21]: I second this

File: stance/predict/overlay.py [hunk 13]
```diff
@@ -30,7 +32,10 @@ def display_overlay(input_video, template_skeletons, benchmark_indices, user_ske
         _, frame = cap.read()
 
         # User skeleton
-        temp = list(user_skeletons[i].vectors.values())
+        try:
+            temp = list(user_skeletons[i].vectors.values())
+        except:
```
Comment (aspadiyath) [inline stance/predict/overlay.py]: Make this an except IndexError

File: stance/motion/squat.py (hunk 11 above)
Comment (aspadiyath) [inline stance/motion/squat.py]: You should delete this too

File: stance/predict/overlay.py [hunk 14]
```diff
@@ -1,8 +1,10 @@
 import cv2
+import matplotlib
```
Comment (aspadiyath) [inline stance/predict/overlay.py]: Delete this as well

**Review by aspadiyath — CHANGES_REQUESTED** [review]
So it looks like you're gonna have a merge conflict, so pull before you push so you can fix them

**Review by aspadiyath — DISMISSED** [review]
Noice noice

File: stance/main.py [hunk 15]
```diff
@@ -59,8 +59,8 @@ def main():
 
     # Overlay the template skeletons on the user's video
     logger.info("Overlaying Video")
-    overlay.display_overlay(args.video_path, motion.template_skeletons)
-    # overlay.compare_user_skeletons(args.video_path, benchmark_zone_indices, user_skeletons, motion)
+    overlay.display_overlay(args.video_path, motion.template_skeletons, benchmark_zone_indices, user_skeletons)
+    # overlay.compare_user_skeletons(args.motion_video, benchmark_zone_indices, user_skeletons, motion)
```
Comment (aspadiyath) [inline stance/main.py]: wait change this back pls

**Review by aspadiyath — CHANGES_REQUESTED** [review]
boi

File: stance/main.py (hunk 15 above)
Comment (jbeta51) [reply stance/main.py]: Which line?
The comment or the method call?

Comment (aspadiyath) [reply stance/main.py]: change motion_video to video_path

**Review by aspadiyath — APPROVED** [review]
J'Approve
